akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolSlot.scala

```diff
@@ -0,0 +1,242 @@
+      case OnNext(x)                      ⇒ throw new IllegalStateException("Unrequested RequestContext: " + x)
+
+      case FromConnection(Request(_)) ⇒
+        inflightRequests = inflightRequests.enqueue(firstRequest)
+        request(totalDemand - remainingRequested)
+        connInport ! OnNext(firstRequest.request)
+        context.become(running(connInport, connOutport))
+
+      case FromConnection(Cancel)     ⇒ if (!isActive) { cancel(); shutdown() } // else ignore and wait for accompanying OnComplete or OnError
+      case FromConnection(OnComplete) ⇒ handleDisconnect(None)
+      case FromConnection(OnError(e)) ⇒ handleDisconnect(Some(e))
+      case FromConnection(OnNext(x))  ⇒ throw new IllegalStateException("Unexpected HttpResponse: " + x)
+    }
+
+    def running(connInport: ActorRef, connOutport: ActorRef): Receive = logMessages("running") {
+      case ev @ (Request(_) | Cancel)     ⇒ connOutport ! ev
+      case ev @ (OnComplete | OnError(_)) ⇒ connInport ! ev
+      case OnNext(rc: RequestContext) ⇒
+        inflightRequests = inflightRequests.enqueue(rc)
+        connInport ! OnNext(rc.request)
+
+      case FromConnection(Request(n)) ⇒ request(n)
+      case FromConnection(Cancel)     ⇒ if (!isActive) { cancel(); shutdown() } // else ignore and wait for accompanying OnComplete or OnError
+
+      case FromConnection(OnNext(response: HttpResponse)) ⇒
+        val requestContext = inflightRequests.head
+        inflightRequests = inflightRequests.tail
+        val delivery = ResponseDelivery(ResponseContext(requestContext, Success(response)))
+        val requestCompleted = SlotEvent.RequestCompleted(slotIx)
+        onNext(delivery :: requestCompleted :: Nil)
+
+      case FromConnection(OnComplete) ⇒ handleDisconnect(None)
+      case FromConnection(OnError(e)) ⇒ handleDisconnect(Some(e))
+    }
+
+    def handleDisconnect(error: Option[Throwable]): Unit = {
+      log.debug("Slot {} disconnected after {}", slotIx, error getOrElse "regular connection close")
+      val results: List[ProcessorOut] = inflightRequests.map { rc ⇒
+        if (rc.retriesLeft == 0) {
+          val reason = error.fold[Throwable](new RuntimeException("Unexpected disconnect"))(identityFunc)
+          ResponseDelivery(ResponseContext(rc, Failure(reason)))
+        } else SlotEvent.RetryRequest(rc.copy(retriesLeft = rc.retriesLeft - 1))
+      }(collection.breakOut)
+      inflightRequests = immutable.Queue.empty
+      onNext(SlotEvent.Disconnected(slotIx, results.size) :: results)
+      if (canceled) onComplete()
+
+      context.become(unconnected)
+    }
+
+    override def onComplete(): Unit = {
+      exposedPublisher.shutdown(None)
+      super.onComplete()
+      shutdown()
+    }
+
+    override def onError(cause: Throwable): Unit = {
+      exposedPublisher.shutdown(Some(cause))
+      super.onError(cause)
+      shutdown()
+    }
+
+    def shutdown(): Unit = {
+      context.become { case _ ⇒ }
```

Comment:
If we are in the `unconnected` state and have a `Cancel` (sent by the downstream) followed by an `OnNext` (sent by the upstream) in the mailbox, we'll end up calling `shutdown()` and `context.stop(self)`. The scaladoc to the latter says that `context.stop` is asynchronous, which means that we might get to handling the `OnNext` before being stopped.
However, the `OnNext` handler in `unconnected` materializes a new `runnableFlow` (which opens a new HTTP connection), that is not automatically stopped/collected when the `SlotProcessor` actor is stopped.

I guess a comment explaining this is warranted here... :)